server/src/main/java/org/candlepin/model/ConsumerCurator.java

```diff
@@ -169,6 +169,68 @@ public Consumer findByVirtUuid(String uuid, String ownerId) {
     }
 
     /**
+     * Lookup all consumers matching the given guest IDs.
+     *
+     * Maps guest ID to the most recent registered consumer that matches it.
+     * Any guest ID not found will not return null.
+     *
+     * If multiple registered consumers report this guest ID (re-registraiton), only the
+     * most recently updated will be returned.
+     *
+     * @param guestId
+     *
+     * @return VirtConsumerMap of guest ID to it's registered guest consumer, or null if
+     * none exists.
+     */
+    @Transactional
+    public VirtConsumerMap getGuestConsumersMap(Owner owner,
+            List<String> guestIds) {
+        VirtConsumerMap guestConsumersMap = new VirtConsumerMap();
+        if (guestIds.size() == 0) {
+            return guestConsumersMap;
+        }
+
+        List<String> possibleGuestIds = Util.getPossibleUuids(guestIds.toArray(
+                new String [guestIds.size()]));
+
+        String sql = "select cp_consumer.uuid from cp_consumer " +
+            "inner join cp_consumer_facts " +
+            "on cp_consumer.id = cp_consumer_facts.cp_consumer_id " +
+            "where cp_consumer_facts.mapkey = 'virt.uuid' and " +
+            "lower(cp_consumer_facts.element) in (:guestids) " +
+            "and cp_consumer.owner_id = :ownerid " +
+            "order by cp_consumer.updated desc";
```

Comment:
Think they should be rewritten? Pretty sure it was by Carter and I suspect there's a good reason but we could try to find out what it was.